Sources/SwiftDocC/Utility/CollectionChanges.swift

```diff
@@ -0,0 +1,219 @@
+            case insert
+        }
+    }
+    
+    /// Creates a new collection changes value from the differences between to collections.
+    ///
+    /// - Parameters:
+    ///   - from: The collection that the base is compared to.
+    ///   - to: The base collection.
+    ///   - areEquivalent: A closure that returns a Boolean value indicating whether two elements are equivalent.
+    init<C>(from: C, to: C, by areEquivalent: (C.Element, C.Element) -> Bool = (==)) where C: BidirectionalCollection, C.Element: Hashable {
+        guard !from.isEmpty else {
+            segments = [.init(kind: .insert, count: to.count)]
+            return
+        }
+        guard !to.isEmpty else {
+            segments = [.init(kind: .remove, count: from.count)]
+            return
+        }
+        
+        var changes = ChangeSegmentBuilder(originalCount: from.count)
+        for change in to.difference(from: from, by: areEquivalent) {
+            switch change {
+            case .remove(let offset, _, _):
+                changes.remove(at: offset)
+            case .insert(let offset, _, _):
+                changes.insert(at: offset)
+            }
+        }
+        segments = changes.segments
+    }
+}
+
+/// A builder that applies collection differences to construct an array of ``Segment`` values.
+///
+/// - Important:
+/// Removals need to be applied in reverse order. All removals need to be applied before applying any insertions. Insertions need to be applied in order.
+private struct ChangeSegmentBuilder {
+    typealias Segment = CollectionChanges.Segment
+    
+    private(set) var segments: [Segment]
+    
+    private var insertStartIndex = 0
+    private var insertStartOffset = 0
+    
+    init(originalCount: Int) {
+        self.segments = [ Segment(kind: .common, count: originalCount) ]
+    }
+    
+    mutating func remove(at removalIndex: Int) {
+        // Removals are applied in reverse order. When the first removal is applied, the only segment is the 'original' count.
+        //
+        // Each removal can be either be at the start of the segment, middle of the segment, or end of the segment.
+        // - After removing from the start of the segment there can be no more removals (since those indices would be in ascending order).
+        // - After removing from the middle, the 'original' segment is split in two with a 'remove' segment in between.
+        //   Since the removal has to be at a lower index, it can only be applied to the split 'original' segment.
+        // - After removing from the end, the 'original' segment is made shorter and a new 'remove' segment is added after it.
+        //   Since the removal has to be at a lower index, it can only be applied to the shortened 'original' segment.
+        //
+        // This process repeats, meaning that every removal is always applied to the first segment.
+        let segment = segments[0]
```

Comment:
Usage of `segment` vs `segments[0]` is inconsistent and slightly confusing to me. Since you need to go through `segments[0]` to mutate anyway, might be worth just using that everywhere to make the code more readable.

Reply:
My intention was that `segment` would indicate a read and `segments[0]` would indicate a write but there are a few places where a subscript is used to read as well. 

I've taken a second look over the insert and remove code and was able to simplify the conditionals a little bit and remove any `segments[0]` reads (but there are still a `segments[1]` read)

Reply:
Updated in https://github.com/apple/swift-docc/pull/420/commits/08346bf76824e7d7aef10aa65d6585c149236b31